File: packages/react-native-autoplay/ios/hybrid/HybridListTemplate.swift
//
//  HybridListTemplate.swift
//  Pods
//
//  Created by Manuel Auer on 15.10.25.
//

import NitroModules

class HybridListTemplate: HybridListTemplateSpec {
    func createListTemplate(config: ListTemplateConfig) throws {
        let template = ListTemplate(config: config)

        try RootModule.withTemplateStore { templateStore in
            templateStore.addTemplate(
                template: template,
                templateId: config.id
            )
        }
    }

    func updateListTemplateSections(
        templateId: String,
        sections: [NitroSection]?
    ) throws -> Promise<Void> {
        return Promise.async {
            try await MainActor.run {
                try RootModule.withAutoPlayTemplate(templateId: templateId) {
                    (template: ListTemplate) in
                    template.updateSections(sections: sections)
                }
            }
        }
    }

    func updateListTemplateContent(
        templateId: String,
        sections: [NitroSection]?,
        detailsHeader: NitroListTemplateDetailsHeader?
    ) throws -> Promise<Void> {
        return Promise.async {
            try await MainActor.run {
                try RootModule.withAutoPlayTemplate(templateId: templateId) {
                    (template: ListTemplate) in
                    template.updateContent(
                        sections: sections,
                        detailsHeader: detailsHeader
                    )
                }
            }
        }
    }

    func updateListTemplatePlayingItem(
        templateId: String,
        itemId: String?
    ) throws -> Promise<Void> {
        return Promise.async {
            try await MainActor.run {
                try RootModule.withAutoPlayTemplate(templateId: templateId) {
                    (template: ListTemplate) in
                    template.updatePlayingItem(itemId: itemId)
                }
            }
        }
    }

    func completeListItemPress(completionId: String) throws -> Promise<Void> {
        return Promise.async {
            ListItemPressCompletionStore.complete(completionId)
        }
    }
}

File: packages/react-native-autoplay/android/src/main/java/com/margelo/nitro/swe/iternio/reactnativeautoplay/HybridListTemplate.kt
package com.margelo.nitro.swe.iternio.reactnativeautoplay

import com.margelo.nitro.core.Promise
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.AndroidAutoTemplate
import com.margelo.nitro.swe.iternio.reactnativeautoplay.template.ListTemplate

class HybridListTemplate : HybridListTemplateSpec() {

    override fun createListTemplate(config: ListTemplateConfig) {
        val context = AndroidAutoSession.getRootContext()
            ?: throw IllegalArgumentException("createListTemplate failed, carContext not found")

        val template = ListTemplate(context, config)
        AndroidAutoTemplate.setTemplate(config.id, template)
    }

    override fun updateListTemplateSections(
        templateId: String, sections: Array<NitroSection>?
    ): Promise<Unit> {
        return Promise.async {
            val template = AndroidAutoTemplate.getTemplate<ListTemplate>(templateId)
            template.updateSections(sections)
        }
    }

    override fun updateListTemplateContent(
        templateId: String,
        sections: Array<NitroSection>?,
        detailsHeader: NitroListTemplateDetailsHeader?
    ): Promise<Unit> {
        return Promise.async {
            val template = AndroidAutoTemplate.getTemplate<ListTemplate>(templateId)
            template.updateSections(sections)
        }
    }

    override fun updateListTemplatePlayingItem(
        templateId: String, itemId: String?
    ): Promise<Unit> {
        return Promise.async {
            Unit
        }
    }

    override fun completeListItemPress(completionId: String): Promise<Unit> {
        return Promise.resolved(Unit)
    }
}
